### PR TITLE
games of players

### PR DESCRIPTION
**Pull request by axel1200**

### CODE REVIEW

File: src/main/java/com/faforever/moderatorclient/api/rest/domain/UserService.java [hunk 1]
```diff
@@ -144,4 +153,32 @@ public BanInfo createBan(@NotNull BanInfo banInfo) {
         banInfo.setAuthor(fafApi.getSelfPlayer());
         return fafApi.post(ElideRouteBuilder.of(BanInfo.class), banInfo);
     }
+
+    public List<GamePlayerStatsFX> getLastHunderedPlayedGamesByFeaturedMod(@NotNull String userId, int page, FeaturedModFX featuredModFX) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/rest/domain/UserService.java]: Please fix the typo. "Hundred"

File: src/main/java/com/faforever/moderatorclient/api/rest/domain/UserService.java [hunk 2]
```diff
@@ -144,4 +153,32 @@ public BanInfo createBan(@NotNull BanInfo banInfo) {
...
+        if (featuredModFX != null) {
+            routeBuilder.filter(ElideRouteBuilder.qBuilder().string("game.featuredMod.technicalName").eq(featuredModFX.getTechnicalName())
+                    .and().string("player.id").eq(userId));
+        } else {
+            routeBuilder.filter(ElideRouteBuilder.qBuilder().string("player.id").eq(userId));
+        }
+        return gamePlayerStatsMapper.map(fafApi.getPage(routeBuilder, 100, page, Collections.emptyMap()));
+    }
+
+    public List<GamePlayerStatsFX> getLastHunderedPlayedGames(@NotNull String userId, int page) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/api/rest/domain/UserService.java]: Please fix the typo. "Hundred"

File: src/main/java/com/faforever/moderatorclient/ui/MainController.java [hunk 3]
```diff
@@ -91,15 +98,24 @@
     // Tab "Recent activity"
     public TableView<Player> userRegistrationFeedTableView;
     public TableView<Teamkill> teamkillFeedTableView;
-
-    public MainController(MapMapper mapMapper, MapVersionMapper mapVersionMapper, PlayerMapper playerMapper, UiService uiService, UserService userService, MapService mapSearchService, AvatarService avatarService) {
+    public TableView<GamePlayerStatsFX> userLastGamesTable;
+    public ChoiceBox<FeaturedModFX> featuredModFilterChoiceBox;
+    public Button loadMoreGamesButton;
+    private Runnable loadMoreGamesRunnable;
+    private int page = 1;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/MainController.java]: Please give it a more meaningful name, like userGamesPage

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 4]
```diff
@@ -83,15 +88,14 @@ static void buildAvatarAssignmentTableView(TableView<AvatarAssignment> tableView
...
-        userIdColumn.setCellFactory(TextFieldTableCell.forTableColumn());
         userIdColumn.setCellValueFactory(param -> new SimpleObjectProperty<>(
                 Optional.ofNullable(param.getValue())
                         .map(avatarAssignment -> avatarAssignment.getPlayer().getId())
                         .orElse(""))
         );
         userIdColumn.setComparator(Comparator.comparingInt(Integer::parseInt));
         userIdColumn.setMinWidth(50);
-        idColumn.setEditable(true);
+        idColumn.setEditable(false);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: This is the default value, so the line can be removed completely.

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 5]
```diff
@@ -459,6 +463,85 @@ protected void updateItem(PlayerFX item, boolean empty) {
...
+
+            {
+                label = new Label();
+                setGraphic(label);
+            }
+            @Override
+            protected void updateItem(OffsetDateTime item, boolean empty) {
+                super.updateItem(item, empty);
+                label.setVisible(!empty);
+                label.setText(item == null ? "unknown Date" : item.format(dateTimeFormatter));
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: date lowercase

File: src/main/java/com/faforever/moderatorclient/ui/domain/GameFX.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+public class GameFX extends AbstractEntityFX {
+    private final StringProperty id;
+    private final StringProperty name;
+    private final ObjectProperty<OffsetDateTime> startTime;
+    private final ObjectProperty<OffsetDateTime> endTime;
+    private final ObjectProperty<Validity> validity;
+    private final ObjectProperty<VictoryCondition> victoryCondition;
+    private final ObservableList<GameReview> reviews;
+    private final ObservableList<GamePlayerStats> playerStats;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/domain/GameFX.java]: To be consistent we should use GameReviewFX and GamePlayerStatsFX objects.

File: src/main/java/com/faforever/moderatorclient/mapstruct/GamePlayerStatsMapper.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.List;
+
+@Mapper(uses = {JavaFXMapper.class, GameMapper.class, PlayerMapper.class})
+public abstract class GamePlayerStatsMapper {
+
+    @Mapping(target = "beforeRating", ignore = true)
+    @Mapping(target = "afterRating", ignore = true)
+    @Mapping(target = "ratingChange", ignore = true)
+    public abstract GamePlayerStatsFX map(GamePlayerStats gamePlayerStats);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/mapstruct/GamePlayerStatsMapper.java]: We need to add the ignores for the backreferences on GameFX object (yes, GameFX is currently not used, but see later comment).

File: src/main/java/com/faforever/moderatorclient/mapstruct/GameMapper.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import com.faforever.moderatorclient.ui.domain.GameFX;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+import java.util.List;
+
+@Mapper(uses = {JavaFXMapper.class, PlayerMapper.class, FeaturedModMapper.class, MapVersionMapper.class})
+public abstract class GameMapper {
+    @Mapping(target = "reviews", ignore = true)
+    public abstract GameFX map(Game game);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/mapstruct/GameMapper.java]: Backreferences for GamePlayerStatsFX are missing (GamePlayerStatsFX usage see later comment)

File: src/main/resources/ui/mainWindow.fxml [hunk 9]
```diff
@@ -129,6 +129,20 @@
...
+                                    <content>
+                                        <VBox>
+                                            <HBox spacing="10">
+                                                <Label text="Show featured Mods:"/>
+                                                <ChoiceBox fx:id="featuredModFilterChoiceBox"/>
+                                                <Button fx:id="loadMoreGamesButton" text="Load more"
+                                                        onAction="#loadMoreGames"/>
+                                            </HBox>
+                                            <TableView fx:id="userLastGamesTable" maxHeight="1.7976931348623157E308"
+                                                       maxWidth="1.7976931348623157E308" VBox.vgrow="ALWAYS"/>
```
Comment (Brutus5000) [inline src/main/resources/ui/mainWindow.fxml]: The maxHeight and maxWidth should be maximum value?

File: src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java [hunk 10]
```diff
@@ -459,6 +463,85 @@ protected void updateItem(PlayerFX item, boolean empty) {
...
+        tableView.getColumns().add(gameIdColum);
+
+        TableColumn<GamePlayerStatsFX, String> gameNameColum = new TableColumn<>("Game Name");
+        gameNameColum.setCellValueFactory(o -> o.getValue().getGame().nameProperty());
+        gameNameColum.setMinWidth(100);
+        tableView.getColumns().add(gameNameColum);
+
+        TableColumn<GamePlayerStatsFX, String> rankedColum = new TableColumn<>("Game Validity");
+        rankedColum.setCellValueFactory(param -> new SimpleObjectProperty<>(
+                param.getValue().getGame().getValidity().name()
```
Comment (Brutus5000) [inline src/main/java/com/faforever/moderatorclient/ui/ViewHelper.java]: This should be a property too (GameFX problem as noted before)

File: pom.xml [hunk 11]
```diff
@@ -83,6 +83,13 @@
             <version>1.6</version>
         </dependency>
 
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-cache</artifactId>
```
Comment (Brutus5000) [inline pom.xml]: Is this actually used?

File: src/main/java/com/faforever/moderatorclient/mapstruct/GameMapper.java (hunk 8 above)
Comment (axel1200) [reply src/main/java/com/faforever/moderatorclient/mapstruct/GameMapper.java]: oh but it did not make any problems hmm

**Pull request merged**